senders/pagerduty/send.go

```diff
@@ -0,0 +1,99 @@
+package pagerduty
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/writeas/go-strip-markdown"
+
+	"github.com/PagerDuty/go-pagerduty"
+
+	"github.com/moira-alert/moira"
+)
+
+const summaryMaxChars = 1024
+
+// SendEvents implements Sender interface Send
+func (sender *Sender) SendEvents(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) error {
+	event := sender.buildEvent(events, contact, trigger, plot, throttled)
+	_, err := pagerduty.ManageEvent(event)
+	if err != nil {
+		return fmt.Errorf("failed to post the event to the pagerduty contact %s : %s. ", contact.Value, err)
+	}
+	return nil
+}
+
+func (sender *Sender) buildEvent(events moira.NotificationEvents, contact moira.ContactData, trigger moira.TriggerData, plot []byte, throttled bool) pagerduty.V2Event {
+	summary := sender.buildSummary(events, trigger, throttled)
+	details := make(map[string]interface{})
+
+	triggerURI := trigger.GetTriggerURI(sender.frontURI)
+	if triggerURI != "" {
+		details["Trigger URI"] = triggerURI
+	} else if trigger.Name != "" {
+		details["Trigger URI"] = trigger.Name
+	}
+
+	if trigger.Desc != "" {
+		details["Description"] = stripmd.Strip(trigger.Desc)
+	}
+
+	var eventList string
+
+	for _, event := range events {
+		line := fmt.Sprintf("\n%s: %s = %s (%s to %s)", event.FormatTimestamp(sender.location), event.Metric, event.GetMetricValue(), event.OldState, event.State)
+		if len(moira.UseString(event.Message)) > 0 {
+			line += fmt.Sprintf(". %s", moira.UseString(event.Message))
+		}
+		eventList += line
+	}
+
+	details["Events"] = eventList
+	if throttled {
+		details["Throttled"] = "Please, fix your system or tune this trigger to generate less events."
+	}
+	payload := &pagerduty.V2Payload{
+		Summary:  summary,
+		Severity: "info",
+		Source:   "moira",
+		Details:  details,
+	}
+	event := pagerduty.V2Event{
+		RoutingKey: contact.Value,
+		Action:     "trigger",
+		Payload:    payload,
+	}
+
+	if len(plot) > 0 && sender.ImageStores[sender.imageStoreID].IsEnabled() {
```

Comment:
I really do not like that you have a lot of places where you use the construction `sender.ImageStores[sender.imageStoreID]`. When writing new code, it can lead to bad consequences. Specifically, here you will fall into a nullPrinterError if the user has given a bad image store name.
Let's do a little differently. In the init method, we will check that such an image store is present in our map (if not, we will write a warning to the log) and if present, we will put in a separate field in Sender structure. Next, when trying to save a picture, we will check if there is any value in this field, if there is, then we will already save the picture.

Reply:
About this comment:

> Let's better make the Enabled flag private and will put it in the Init method after configuration.

Let's then set it up in a separate field if image store IsEnabled.